spec/deduplication.md

```diff
@@ -0,0 +1,183 @@
+- **Immutability**: Once stored, content cannot be modified
+- **Deduplication**: Identical content is automatically deduplicated at the storage level
+
+## Deduplication Procedure
+
+### 1. File Processing and Chunking
+
+When a file is processed for upload, it undergoes the following steps:
+
+```mermaid
+graph TD
+    A[File Input] --> B[Content-Defined Chunking]
+    B --> C[Hash Computation]
+    C --> D[Chunk Creation]
+    D --> E[Deduplication Query]
+```
+
+1. **File Reading**: Files are read in blocks (default 8MB) to manage memory usage
+2. **Chunking**: Content-defined chunking using GearHash algorithm creates variable-sized chunks
+3. **Hash Computation**: Each chunk's content is hashed using a cryptographic hash function (Blake3-based MerkleHash)
+4. **Chunk Object Creation**: Chunks are wrapped with metadata including hash, size, and data
+
+### 2. Multi-Level Deduplication Strategy
+
+Xet employs a three-tiered deduplication strategy to maximize efficiency while minimizing latency:
+
+#### Level 1: Local Session Deduplication
+
+**Scope**: Current upload session
+**Mechanism**: In-memory hash lookup table
+**Purpose**: Eliminate redundancy within the current file or session
+
+**Benefits**:
+
+- Fastest lookup (in-memory)
+- Zero network overhead
+- Immediate deduplication feedback
+
+#### Level 2: Cached Metadata Deduplication
+
+**Scope**: Previously uploaded files and sessions
+**Mechanism**: Local shard file metadata cache
+**Purpose**: Leverage deduplication against recently uploaded content
+
+**Benefits**:
+
+- Fast local disk access
+- No network latency
+- Persistent across sessions
+
+#### Level 3: Global Deduplication API
+
+**Scope**: Entire Xet ecosystem
+**Mechanism**: Global deduplication service with HMAC protection
+**Purpose**: Discover deduplication opportunities across all users and repositories
+
+### 3. Global Deduplication Process
+
+The global deduplication system provides deduplication capabilities across all data that is managed by the xet system:
+
+#### Eligibility Criteria
+
+Not all chunks are eligible for global deduplication queries to manage system load:
+
+1. **First chunk**: The first chunk of every file is always eligible.
+2. **Hash pattern matching**: Chunks are eligible if: the last 8 bytes of the hash interpreted as a little-endian 64 bit integer % 1024 == 0.
+3. **Spacing constraints**: There must be a minimum of 4MB between global deduplication queries within a file.
+
+#### Query Process
+
+1. **Background Query**: Global deduplication queries run asynchronously to avoid blocking upload
+2. **HMAC Protection**: Chunk hashes are protected using repository-specific HMAC keys
+3. **Shard Response**: When a match is found, the API returns a shard containing:
+   - **CAS Info Section**: Contains metadata about many xorbs that store chunks
+   - **HMAC Key**: Included in the shard metadata header for decryption purposes
```

Comment:
Is this supposed to be for *encryption* purposes? Step 5 indicates that this key is used for encrypting the local chunk hashes to compare.

Reply:
yeah, just pointing out that the line says `decryption` instead of `encryption`.